cli/tests/pcluster/utils/test_pcluster_utils.py

```diff
@@ -0,0 +1,229 @@
+    boto3_stubber("cloudformation", mocked_requests, generate_errors=True)
+    expected_message = "Unable to update stack template for stack {stack_name}: {emsg}".format(
+        stack_name=stack_name, emsg=emsg
+    )
+    with pytest.raises(SystemExit, match=expected_message):
+        utils.update_stack_template(stack_name, template_body, cfn_params)
+
+
+def test_get_cluster_substacks(mocker):
+    """Verify that utils.get_cluster_substacks returns the correct stacks."""
+    stack_type = "AWS::CloudFormation::Stack"
+    resources = [
+        {"ResourceType": "Not a stack", "ResourceName": "name_one"},
+        {"ResourceType": stack_type, "ResourceName": "name_two"},
+        {"ResourceType": "Also not a stack", "ResourceName": "name_three"},
+        {"ResourceType": stack_type, "ResourceName": "name_four"},
+    ]
+    mocker.patch("pcluster.utils.get_stack_resources").return_value = resources
+    cluster_name = "cluster_name"
+    stack_name = utils.get_stack_name(cluster_name)
+    expected_substacks = [d for d in resources if d.get("ResourceType") == stack_type]
+    observed_substacks = utils.get_cluster_substacks(cluster_name)
+    utils.get_stack_resources.assert_called_with(stack_name)
+    assert_that(observed_substacks).is_equal_to(expected_substacks)
+
+
+def test_get_cloudwatch_logs_substack(mocker):
+    """Verify that utils.get_cloudwatch_logs_substack returns the expected stack."""
+    cluster_name = "cluster_name"
+    cw_logs_substack = {
+        "StackName": "{stack_name}-{substack_prefix}-".format(
+            stack_name=utils.get_stack_name(cluster_name), substack_prefix=utils.CW_LOGS_SUBSTACK_PREFIX
+        ),
+    }
+    substacks_response = [
+        {"StackName": "substack_one"},
+        {"StackName": "substack_two"},
+        {"StackName": "not_a_substack"},
+        cw_logs_substack,
+    ]
+    mocker.patch("pcluster.utils.get_cluster_substacks").return_value = substacks_response
+    observed_substack = utils.get_cloudwatch_logs_substack(cluster_name)
+    utils.get_cluster_substacks.assert_called_with(cluster_name)
+    assert_that(observed_substack).is_equal_to(cw_logs_substack)
+
+
+def test_get_cloudwatch_logs_substack_no_match(mocker):
+    """Verify that utils.get_cloudwatch_logs_substack raises the expected exception when it can't find the substack."""
+    cluster_name = "cluster_name"
+    substacks_response = [{"StackName": "substack_one"}, {"StackName": "substack_two"}, {"StackName": "not_a_substack"}]
+    mocker.patch("pcluster.utils.get_cluster_substacks").return_value = substacks_response
+    expected_message = "Unable to get CloudWatch logs substack for cluster {}".format(cluster_name)
+    with pytest.raises(SystemExit, match=expected_message) as sysexit:
+        utils.get_cloudwatch_logs_substack(cluster_name)
+    assert_that(sysexit.value.code).is_not_equal_to(0)
+    utils.get_cluster_substacks.assert_called_with(cluster_name)
+
+
+def test_stack_exists_false(boto3_stubber):
```

Comment:
We can parametrize this test and the following one.
The parameters could be the describe_stacks_response and the returned value (true/false)

Reply:
I agree that these tests should be refactored to avoid duplication. I am going to do it today after addressing the change Francesco asked about because I think that having all of the cluster's log groups persist rather than just the one that is currently will change how I go about it (and test it).

Reply:
Not in the case of this specific function, though.